src/python/pants/engine/target_test.py

```diff
@@ -0,0 +1,191 @@
+    with pytest.raises(ValueError) as exc:
+        HaskellTarget({"invalid_field": True})
+    assert "Unrecognized field `invalid_field=True` for target type `haskell`." in str(exc)
+
+
+def test_get_primitive_field() -> None:
+    extensions = ["GhcExistentialQuantification"]
+    extensions_field = HaskellTarget({HaskellGhcExtensions.alias: extensions}).get(
+        HaskellGhcExtensions
+    )
+    assert extensions_field.raw_value == extensions
+    assert extensions_field.value == extensions
+
+    default_extensions_field = HaskellTarget({}).get(HaskellGhcExtensions)
+    assert default_extensions_field.raw_value is None
+    assert default_extensions_field.value == []
+
+
+def test_get_async_field() -> None:
+    def hydrate_field(
+        *, raw_source_files: List[str], hydrated_source_files: Tuple[str, ...]
+    ) -> HaskellSourcesResult:
+        sources_field = HaskellTarget({HaskellSources.alias: raw_source_files}).get(HaskellSources)
+        assert sources_field.raw_value == raw_source_files
+        result: HaskellSourcesResult = run_rule(
+            hydrate_haskell_sources,
+            rule_args=[sources_field],
+            mock_gets=[
+                MockGet(
+                    product_type=Snapshot,
+                    subject_type=PathGlobs,
+                    mock=lambda _: Snapshot(
+                        directory_digest=EMPTY_DIRECTORY_DIGEST,
+                        files=hydrated_source_files,
+                        dirs=(),
+                    ),
+                )
+            ],
+        )
+        return result
+
+    # Normal field
+    expected_files = ("monad.hs", "abstract_art.hs", "abstract_algebra.hs")
+    assert (
+        hydrate_field(
+            raw_source_files=["monad.hs", "abstract_*.hs"], hydrated_source_files=expected_files
+        ).snapshot.files
+        == expected_files
+    )
+
+    # Test pre-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=[0, 1, 2], hydrated_source_files=())  # type: ignore[call-arg]
+    assert "Not all elements of the iterable have type" in str(exc)
+
+    # Test post-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=["*.js"], hydrated_source_files=("not_haskell.js",))
+    assert "Received non-Haskell sources" in str(exc)
+
+
+def test_has_fields() -> None:
+    class UnrelatedField(PrimitiveField):
+        alias: ClassVar = "unrelated"
+        raw_value: Optional[bool]
+
+        @memoized_property
+        def value(self) -> bool:
+            if self.raw_value is None:
+                return False
+            return self.raw_value
+
+    tgt = HaskellTarget({})
+    assert tgt.has_fields([]) is True
+    assert tgt.has_fields([HaskellGhcExtensions]) is True
+    assert tgt.has_fields([UnrelatedField]) is False
+    assert tgt.has_fields([HaskellGhcExtensions, UnrelatedField]) is False
+
+
+def test_field_hydration_is_lazy() -> None:
```

Comment:
Idle question: How would you feel about a `./pants validate-build-files` goal or similar? People used to use `./pants list` to make sure BUILD files mostly look right, and it feels like something to force pre- and post-hydration validation, without all of the cost of actually compiling and running tests, would be useful.

Reply:
I'd love to have that!